Wait for complete download in core life start

martinfurmanski opened the pull request:
Previously we would wait just for the snapshot to have been installed
and potentially double-start local database.

martinfurmanski committed:
Wait for complete download in core life start

Previously we would wait just for the snapshot to have been installed
and potentially double-start local database.

lutovich commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java:
```diff
@@ -86,6 +86,7 @@ public void run()
         }
         catch ( InterruptedException e )
         {
+            Thread.interrupted();
```
I think this line will to nothing. It should be `Thread.currentThread().interrupt()` if intention is to restore the interruption status.

martinfurmanski committed:
Cleaner error handling around store copy

* Fatal and unknown exceptions lead to a panic.
  Typically these are life-cycle exceptions, I/O exceptions, etc.

* StoreCopyFailedException is now a proper wrapper for benign issues.
  The unnecessary StreamingTransactionsFailed exception was thus removed,
  since it could be subsumed by this in all places. Ideally we would
  get rid of this exception completely.

* The CoreStateDownloader now avoids exceptions for regular control flow,
  and exceptions are just used for exceptional circumstances. Specifically,
  the StoreCopyFailedException is not a valid outcome. Either an exception
  with known failure handling semantics is thrown, or a boolean is returned
  indicating if the operation succeeded or needs to be retried.

martinfurmanski committed:
Replace random member with same role member in stress test

Replacing with any role is problematic, since you can completely
kill off the intended quorum and end up with < 2 cores, as an example.

RagnarW commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/CoreStateDownloader.java:
```diff
@@ -71,97 +76,160 @@ public CoreStateDownloader( LocalDatabase localDatabase, Lifecycle startStopOnSt
-            {
-                throw new StoreCopyFailedException( "StoreId mismatch and not empty" );
-            }
+        if ( !isEmptyStore && !remoteStoreId.equals( localDatabase.storeId() ) )
```
We check store id in any request related to catchup. So I do not thing that this check i necessary. Especially now when we use an AddressProvider we comparing against an adress we might not even use later.

Possible we should add store id into to CatchuRequest interface so that we force it on any request that involves catchup

RagnarW replied:
However this is not really related to the changes in this PR

martinfurmanski replied:
StoreId isn't included actually on the subsequent CoreSnapshotRequest.

RagnarW commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/CoreStateDownloader.java:
```diff
@@ -71,97 +76,160 @@ public CoreStateDownloader( LocalDatabase localDatabase, Lifecycle startStopOnSt
+
+        /* We install the snapshot after the store has been downloaded,
+         * so that we are not left with a state ahead of the store. */
+        snapshotService.installSnapshot( coreSnapshot );
```
are we okay with throwing whatever here?

martinfurmanski replied:
It only declares I/O exceptions, and I think such should be thrown and be handled fatally.
